collector/qdisc_linux.go

```diff
@@ -0,0 +1,296 @@
+	TCA_RATE
+	TCA_FCNT
+	TCA_STATS2
+	TCA_STAB
+	__TCA_MAX
+)
+
+const (
+	TCA_STATS_UNSPEC = iota
+	TCA_STATS_BASIC
+	TCA_STATS_RATE_EST
+	TCA_STATS_QUEUE
+	TCA_STATS_APP
+	TCA_STATS_RATE_EST64
+	__TCA_STATS_MAX
+)
+
+// See struct tc_stats in /usr/include/linux/pkt_sched.h
+type TC_Stats struct {
+	Bytes      uint64
+	Packets    uint32
+	Drops      uint32
+	Overlimits uint32
+	Bps        uint32
+	Pps        uint32
+	Qlen       uint32
+	Backlog    uint32
+}
+
+// See /usr/include/linux/gen_stats.h
+type TC_Stats2 struct {
+	// struct gnet_stats_basic
+	Bytes   uint64
+	Packets uint32
+	// struct gnet_stats_queue
+	Qlen       uint32
+	Backlog    uint32
+	Drops      uint32
+	Requeues   uint32
+	Overlimits uint32
+}
+
+type Metric struct {
+	IfaceIdx   uint32
+	Parent     uint32
+	Handle     uint32
+	Kind       string
+	Bytes      uint64
+	Packets    uint32
+	Drops      uint32
+	Requeues   uint32
+	Overlimits uint32
+}
+
+// See if_indextoname(3)
+func ifIndexToName(index uint32) string {
+	var name string
+	cName := C.CString(name)
+	defer C.free(unsafe.Pointer(cName))
+
+	ret := C.if_indextoname(C.uint(index), cName)
+	return C.GoString(ret)
+}
+
+func parseTCAStats(attr netlink.Attribute) TC_Stats {
+	var stats TC_Stats
+	stats.Bytes = nlenc.Uint64(attr.Data[0:8])
+	stats.Packets = nlenc.Uint32(attr.Data[8:12])
+	stats.Drops = nlenc.Uint32(attr.Data[12:16])
+	stats.Overlimits = nlenc.Uint32(attr.Data[16:20])
+	stats.Bps = nlenc.Uint32(attr.Data[20:24])
+	stats.Pps = nlenc.Uint32(attr.Data[24:28])
+	stats.Qlen = nlenc.Uint32(attr.Data[28:32])
+	stats.Backlog = nlenc.Uint32(attr.Data[32:36])
+	return stats
+}
+
+func parseTCAStats2(attr netlink.Attribute) TC_Stats2 {
+	var stats TC_Stats2
+
+	nested, _ := netlink.UnmarshalAttributes(attr.Data)
+
+	for _, a := range nested {
+		switch a.Type {
+		case TCA_STATS_BASIC:
+			stats.Bytes = nlenc.Uint64(a.Data[0:8])
+			stats.Packets = nlenc.Uint32(a.Data[8:12])
+		case TCA_STATS_QUEUE:
+			stats.Qlen = nlenc.Uint32(a.Data[0:4])
+			stats.Backlog = nlenc.Uint32(a.Data[4:8])
+			stats.Drops = nlenc.Uint32(a.Data[8:12])
+			stats.Requeues = nlenc.Uint32(a.Data[12:16])
+			stats.Overlimits = nlenc.Uint32(a.Data[16:20])
+		default:
+			// TODO: TCA_STATS_APP
+		}
+	}
+
+	return stats
+}
+
+func getQdiscMsgs() ([]netlink.Message, error) {
+	const familyRoute = 0
+
+	c, err := netlink.Dial(familyRoute, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial netlink: %v", err)
+	}
+	defer c.Close()
+
+	req := netlink.Message{
+		Header: netlink.Header{
+			Flags: netlink.HeaderFlagsRequest | netlink.HeaderFlagsDump,
+			Type:  38, // RTM_GETQDISC
+		},
+		Data: []byte{0},
+	}
+
+	// Perform a request, receive replies, and validate the replies
+	msgs, err := c.Execute(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute request: %v", err)
+	}
+
+	return msgs, nil
+}
+
+// See https://tools.ietf.org/html/rfc3549#section-3.1.3
+func parseMessage(msg netlink.Message) (Metric, error) {
```

Comment:
Since you're handling netlink messages and the interesting cases that come with that, it might be worth making a separate package like `github.com/ema/qdisc` and then importing that for `node_exporter`.  I like to promote code reusability wherever possible.